online/src/app/product-details/product-details.component.ts

```diff
@@ -14,31 +16,34 @@ import { admin } from '../utils';
   styleUrls: ['./product-details.component.scss']
 })
 export class ProductDetailsComponent implements OnInit {
-  product: Product | undefined;
+  id: number | undefined;
   products: ProductViewModel[] | undefined;
-  hasAdminRole = this.authService.hasRoleType(admin);
+  roleSelector = this.store.select(adminRoleSelector);
+  hasAdminRole: boolean | undefined;
+  product: Product | undefined;
+  selectedProduct$ = this.store.select(selectOneProduct);
 
-  constructor(private productService: ProductService, private location: Location, private route: ActivatedRoute, private authService: AuthentificationService) {
 
-  }
+  constructor(private location: Location,
+    private route: ActivatedRoute,
+    private store: Store<AppState>) {
 
-  ngOnInit(): void {
-    const id = this.route.snapshot.params['id'];
-    this.getProductById(id).subscribe((item: Product) => this.product = item);
   }
 
-  getProductById(id: number): Observable<Product> {
-    return this.productService.getProductById(id);
+  ngOnInit(): void {
+    this.id = this.route.snapshot.params['id'];
+    if (this.id) {
+      this.store.dispatch(getProduct({ id: this.id }));
+    }
+    this.roleSelector.subscribe((role) => { this.hasAdminRole = role });
```

Comment:
Generally, prefer to use the observable directly without reassigning into another static variable. You have the async pipe for the template and for the ts you can pipe it or use withLatestFrom (prolly easier) which returns a promise.

This is because the way you have it here, it's prone to memory leaks and you normally would have to close the subscription on something like ngOnDestroy lifecycle event.

Reply:
I handled the memory leak by adding the implementation of OnDestroy and unsubscribing the subscription of the role in ngOnDestroy event.